unit-test-security-rules-v9/test/database.spec.mjs

```diff
@@ -0,0 +1,140 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import assert from 'assert';
+import { readFileSync, createWriteStream } from 'fs';
+import http from 'http';
+
+// Workarounds for npm not recognizing named exports in cjs.
+import appNs from 'firebase/app';
+const { initializeApp, deleteApp, getApps, setLogLevel } = appNs;
+import dbNs from 'firebase/database';
+const { getDatabase, useDatabaseEmulator, ref, get, update } = dbNs;
+
+import adminAppNs from 'firebase-admin/app';
+const { initializeApp: initializeAdminApp } = adminAppNs;
+import adminDatabaseNs from 'firebase-admin/database';
+const { getDatabase: getAdminDatabase } = adminDatabaseNs;
+
+/**
+ * The emulator will accept any database name for testing.
+ */
+const DATABASE_NAME = "demo-database-emulator";
+
+/**
+ * The FIREBASE_DATABASE_EMULATOR_HOST environment variable is set automatically
+ * by "firebase emulators:exec"
+ */
+const { FIREBASE_DATABASE_EMULATOR_HOST } = process.env;
+if (!FIREBASE_DATABASE_EMULATOR_HOST) {
+  throw new Error('Missing environment variable FIREBASE_DATABASE_EMULATOR_HOST. Consider running tests via `firebase emulators:exec "npm test"`.');
+}
+
+const DATABASE_URL = `http://${FIREBASE_DATABASE_EMULATOR_HOST}/?ns=${DATABASE_NAME}`
+const { hostname: DATABASE_HOST, port: DATABASE_PORT } = new URL(DATABASE_URL);
+
+const COVERAGE_URL = `http://${FIREBASE_DATABASE_EMULATOR_HOST}/.inspect/coverage?ns=${DATABASE_NAME}`;
+
+before(async () => {
+  initializeAdminApp({ databaseURL: DATABASE_URL });
+
+  // Set database rules before running these tests.
+  await getAdminDatabase().setRules(readFileSync("database.rules.json", "utf8"));
+
+  // Silence expected rules rejections from Firestore SDK. Unexpected rejections
+  // will still bubble up and will be thrown as an error (failing the tests).
+  setLogLevel('error');
+});
+
+after(async () => {
+  // Delete all the FirebaseApp instances created during testing.
+  // Note: this does not affect or clear any data.
+  await Promise.all(getApps().map(deleteApp));
+
+  // Write the coverage report to a file
+  const coverageFile = 'database-coverage.html';
+  const fstream = createWriteStream(coverageFile);
+  await new Promise((resolve, reject) => {
+      http.get(COVERAGE_URL, (res) => {
+        res.pipe(fstream, { end: true });
+
+        res.on("end", resolve);
+        res.on("error", reject);
+      });
+  });
+
+  console.log(`View database rule coverage information at ${coverageFile}\n`);
+});
+
+/**
+ * Creates a new client FirebaseApp with authentication and returns the Database instance.
+ */
+function getAuthedDatabase(auth) {
+  const app = initializeApp(
+    { databaseURL: DATABASE_URL },
+    'demo-app-' + Date.now() + Math.random()
```

Comment:
Is the `Math.random()` in case this runs more than once in 1ms?

Reply:
Yup